### PR TITLE
feat(spanner): CommitStats in CommitResponse

### PR DESCRIPTION
**Pull request by jiren**
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-cloud-ruby/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea.
- [x] Follow the instructions in [CONTRIBUTING](CONTRIBUTING.md). Most importantly, ensure the tests and linter pass by running `bundle exec rake ci` in the gem subdirectory.
- [x] Update code documentation if necessary.

closes: #8057

### CODE REVIEW

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (skuruppu) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb:818]: I'm curious, wouldn't this be a breaking change for users? If they are currently calling `upsert`, they are expecting it to return a timestamp. But now we return a `CommitResponse` so wouldn't that throw an error?

E.g. the following code will break:
```
ts = db.upsert "users", [{ id:1, name: "Charlie", active: false }, ...]
# use ts for something else
...
```

File: google-cloud-spanner/lib/google/cloud/spanner/commit_response/commit_stats.rb
Comment (skuruppu) [inline google-cloud-spanner/lib/google/cloud/spanner/commit_response/commit_stats.rb:29]:
```suggestion
          # @private Creates a new CommitStats instance.
```

File: google-cloud-spanner/lib/google/cloud/spanner/commit_response/commit_stats.rb
Comment (skuruppu) [inline google-cloud-spanner/lib/google/cloud/spanner/commit_response/commit_stats.rb:48]:
```suggestion
          # Creates a new CommitStats instance from a
```

File: google-cloud-spanner/lib/google/cloud/spanner/session.rb
Comment (skuruppu) [inline google-cloud-spanner/lib/google/cloud/spanner/session.rb:507]: Same here regarding this being a breaking change.

**Review by skuruppu — COMMENTED** [review]
@jiren I have a concern that this implementation might be causing a breaking change for users who were previously reading the commit timestamp that's returned by most of the methods that are modified. We avoided doing this in most of the other language implementations. Let me know if I misunderstood. But if we could come up with a way to make this a non-breaking change, that would be ideal.

**Comment by jiren** [comment]
> @jiren I have a concern that this implementation might be causing a breaking change for users who were previously reading the commit timestamp that's returned by most of the methods that are modified. We avoided doing this in most of the other language implementations. Let me know if I misunderstood. But if we could come up with a way to make this a non-breaking change, that would be ideal.

Yes, In this case I am planning to extend(delegate in ruby [ref](https://github.com/googleapis/google-cloud-ruby/blob/8fd12a741a3c99a2b42e19d966fbeb2587c35c9c/google-cloud-spanner/lib/google/cloud/spanner/instance/list.rb#L25)) wrapper CommitResponse from Timestamp then it will works. So commit response works as a timestamp with stats attributes.

```ruby
class CommitResponse < DelegateClass(::Time); end
```

**Comment by skuruppu** [comment]
> > @jiren I have a concern that this implementation might be causing a breaking change for users who were previously reading the commit timestamp that's returned by most of the methods that are modified. We avoided doing this in most of the other language implementations. Let me know if I misunderstood. But if we could come up with a way to make this a non-breaking change, that would be ideal.
> 
> Yes, In this case I am planning to extend(delegate in ruby [ref](https://github.com/googleapis/google-cloud-ruby/blob/8fd12a741a3c99a2b42e19d966fbeb2587c35c9c/google-cloud-spanner/lib/google/cloud/spanner/instance/list.rb#L25)) wrapper CommitResponse from Timestamp then it will works. So commit response works as a timestamp with stats attributes.
> 
> ```ruby
> class CommitResponse < DelegateClass(Time); end
> ```

Oh cool, so does that means that users who are already using these APIs will not have to change their code?

**Comment by hengfengli** [comment]
@jiren I am not keen to use `DelegateClass` to let `CommitResponse` inherit from `Time`. This confuses me because `timestamp` should be a field of `CommitResponse`. The relationship here does not look good to me, although it may solve the backward-compatible issue.

**Comment by skuruppu** [comment]
Just to clarify the requirements @jiren, this feature is only meant to be used for debugging queries. So we only expect people to use the flag in development. So we don't want users to have to rewrite their production applications.

In other languages, we added extra methods like `insertWithOptions()` that returns a `CommitResponse`. We might have to do the same here.

File: google-cloud-spanner/test/google/cloud/spanner/client/transaction_test.rb
Comment (hengfengli) [inline google-cloud-spanner/test/google/cloud/spanner/client/transaction_test.rb:481]: It's unclear why we add this test. We only need to test the commit stats feature.

File: google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb
Comment (hengfengli) [inline google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb:163]: Duplicated code?

**Comment by jiren** [comment]
> @jiren I am not keen to use `DelegateClass` to let `CommitResponse` inherit from `Time`. This confuses me because `timestamp` should be a field of `CommitResponse`. The relationship here does not look good to me, although it may solve the backward-compatible issue.
> 
> Instead of changing the return type of existing methods, can we add new methods to return `CommitResponse`? This is what we have done in other languages.

@hengfengli One thing we can do in case we go with the inheritance - add more detail in the documentation 
For the second option adding additional methods, I think it will add a lot more redundant code.
I have one more option return `Time` or `CommitResponse` based on `commit_stats` method argument.

i.e
```
def insert table, rows, commit_stats: nil
   ...
   ...
   if commit_stats 
     return commit_response
   else
     return commit_response.timestamp.
   end
end
```

File: google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb
Comment (hengfengli) [inline google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb:157]: Duplicated code?

File: google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb
Comment (hengfengli) [inline google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb:129]: Why do we test `commit` twice here?

File: google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb
Comment (hengfengli) [inline google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb:344]: Same here. Why do we test `commit` twice?

File: google-cloud-spanner/test/google/cloud/spanner/session/commit_test.rb
Comment (hengfengli) [inline google-cloud-spanner/test/google/cloud/spanner/session/commit_test.rb:94]: I feel that this adds a lot of redundant test code. Can we try to avoid these changes?

**Review by hengfengli — CHANGES_REQUESTED** [review]
Thanks for making the changes. There are two major issues: 

1. A breaking change. We need to avoid this. One idea I have is that if the user specifies an `commit_options` then we return `commit_response` instead of `timestamp`. If `commit_options` is not given, then we return `timestamp`. Otherwise, we need to add new methods to just return `commit_response` (this would add a lot of new methods). 
2. Too many redundant test code. They are most duplicated code. Please try to simplify the new added test code. Only test what we just add and avoid too many changes, which is hard for us to review the code.

File: google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb:1335]: Ok, but unit test will fail on CI.

**Comment by jiren** [comment]
@hengfengli I am adding option for `commit_options`  and based on that I going to return timestamp or stats.

File: google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb
Comment (hengfengli) [reply google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb:1335]: Usually, we run unit tests locally before committing to a PR.

File: google-cloud-spanner/acceptance/spanner_helper.rb
Comment (jiren) [reply google-cloud-spanner/acceptance/spanner_helper.rb:39]: Ok, removing it.

**Comment by hengfengli** [comment]
> @hengfengli I am adding option for `commit_options` and based on that I going to return timestamp or stats.

This sounds good to me.

File: google-cloud-spanner/test/google/cloud/spanner/session/commit_test.rb
Comment (jiren) [reply google-cloud-spanner/test/google/cloud/spanner/session/commit_test.rb:94]: Yes working on redundant test case code.

File: google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb:1335]: Ok, I am removing it these auto generated files. While running on local you need to generate Gapic client to pickup changes from googleapis [branch](https://github.com/googleapis/googleapis/tree/spanner_commit_stats). 
FYI Tests on Kokoro CI will fail.

Comment (hengfengli) [reply google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb:1335]: Yea, that's fine. Once the generated code is merged into master, we rebase this branch and it will pass.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Updated for backward compatibility.

File: google-cloud-spanner/lib/google/cloud/spanner/commit_response/commit_stats.rb
Comment (jiren) [reply google-cloud-spanner/lib/google/cloud/spanner/commit_response/commit_stats.rb:49]: Done

File: google-cloud-spanner/acceptance/spanner/client/batch_update_test.rb
Comment (jiren) [reply google-cloud-spanner/acceptance/spanner/client/batch_update_test.rb]: Updated for backward compatibility.

File: google-cloud-spanner/acceptance/spanner/client/transaction_test.rb
Comment (jiren) [reply google-cloud-spanner/acceptance/spanner/client/transaction_test.rb]: added separate test for simple transaction and removed above test.

File: google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb
Comment (jiren) [reply google-cloud-spanner/test/google/cloud/spanner/client/commit_test.rb]: removed. Added to test rows and single row.

**Comment by jiren** [comment]
@hengfengli @skuruppu Updated for backward compatibility. 
- Added `commit_options : { return commit_stats: } `
- Condition based response - If user provides commit options then return `CommitResponse` object else `Time` obj.

File: google-cloud-spanner/acceptance/spanner/client/transaction_test.rb
Comment (hengfengli) [inline google-cloud-spanner/acceptance/spanner/client/transaction_test.rb:206]: Do we need to add query options test?

File: google-cloud-spanner/lib/google/cloud/spanner/commit_response/commit_stats.rb
Comment (hengfengli) [reply google-cloud-spanner/lib/google/cloud/spanner/commit_response/commit_stats.rb:49]: @jiren It seems that this is still not fixed.

**Comment by hengfengli** [comment]
@skuruppu It now looks good to me. Can you please take a quick look?

File: google-cloud-spanner/acceptance/spanner/client/transaction_test.rb
Comment (jiren) [reply google-cloud-spanner/acceptance/spanner/client/transaction_test.rb]: removed.

**Review by skuruppu — APPROVED** [review]
I think this is reasonable if it's normal to alter the return value in ruby based on inputs. Thanks for making this a non-breaking feature.

**Comment by skuruppu** [comment]
@jiren I have a small request to make. Would you be able to remove any hand-written support for `overload_delay`? This proto field is no longer getting released so when the protos are published, it will no longer be there. I hope it's not too much trouble.

**Comment by jiren** [comment]
> @jiren I have a small request to make. Would you be able to remove any hand-written support for `overload_delay`? This proto field is no longer getting released so when the protos are published, it will no longer be there. I hope it's not too much trouble.

Working on it.

**Comment by jiren** [comment]
> @jiren I have a small request to make. Would you be able to remove any hand-written support for `overload_delay`? This proto field is no longer getting released so when the protos are published, it will no longer be there. I hope it's not too much trouble.

@skuruppu Removed overload_delay code.

**Review by hengfengli — APPROVED** [review]

**Comment by skuruppu** [comment]
@jiren the generated code has been merged so please rebase this PR and get it ready for merging.

**Comment by jiren** [comment]
> @jiren the generated code has been merged so please rebase this PR and get it ready for merging.

Done

**Comment by skuruppu** [comment]
@dazuma or @quartzmo, we would really like to get an approval from one of you so that we can merge in this PR.

**Pull request merged**
